src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -145,23 +222,41 @@ protected Object decode(
             position.setNetwork(new Network(CellTower.from(mcc, mnc, lac, cid)));
         }
 
+        String unused = parser.next();
+
         String checksum = parser.next();
 
         if (channel != null) {
-            if (event.equals("3")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,d*31\r\n", remoteAddress));
-            } else if (event.equals("X") || event.equals("4")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,x*2D\r\n", remoteAddress));
-            } else if (event.equals("Z")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,z*2F\r\n", remoteAddress));
-            } else if (Character.isLowerCase(status.charAt(0))) {
-                String response = "$EAVACK," + event + "," + checksum;
-                response += Checksum.nmea(response) + "\r\n";
-                channel.writeAndFlush(new NetworkMessage(response, remoteAddress));
-            }
+
+            sendAcknowledge(status, event, checksum, channel, remoteAddress);
+
+            String devicePassword = Context.getIdentityManager()
+                .getDevicePassword(deviceSession.getDeviceId(), DEFAULT_DEVICE_PASSWORD);
+            sendEventResponse(event, devicePassword, channel, remoteAddress);
         }
 
         return position;
     }
 
+    @Override
+    protected Object decode(
+        Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        String sentence = (String) msg;
+
+        if (sentence.startsWith("$ECHK")) {
+            Parser parser = new Parser(PATTERN_ECHK, sentence);
```

Comment:
This can probably be moved to corresponding methods. Also instead of `handleEchk` we usually names like `decodeEchk`.

Reply:
Done.